form/form.go

```diff
@@ -0,0 +1,279 @@
+	"strconv"
+	"strings"
+)
+
+const tagName = "form"
+
+type formOptions struct {
+	// IndexedArray indicates that contrary to standard "Rack-style" form
+	// encoding, array items should be index like `arr[0]=...&arr[1]=...`
+	// (normally it'd be `arr[]=...`).
+	IndexedArray bool
+
+	// Empty indicates that a field's value should be emptied in that its value
+	// should be an empty string. It's used to workaround the fact that an
+	// empty string is a string's zero value and wouldn't normally be encoded.
+	Empty bool
+
+	// Invert indicates that a boolean field's value should be inverted. False
+	// is the zero value for a boolean so it's convention in the library to
+	// specify a `No*` field to allow a false to be passed to the API. These
+	// fields should be annotated with `invert`.
+	Invert bool
+
+	// Zero indicates a field that's specifically defined to workaround the
+	// fact because 0 is the "zero value" of all int/float types, we can't
+	// properly encode an explicit 0. It indicates that an explicit zero should
+	// be sent.
+	Zero bool
+}
+
+func AppendTo(values *RequestValues, i interface{}) {
+	reflectValue(values, reflect.ValueOf(i), nil, nil)
+}
+
+func formatName(names []string) string {
+	if len(names) < 1 {
+		panic("Not allowed 0-length names slice")
+	}
+
+	fullName := names[0]
+	for i := 1; i < len(names); i++ {
+		fullName += "[" + names[i] + "]"
+	}
+	return fullName
+}
+
+func reflectValue(values *RequestValues, val reflect.Value, names []string, options *formOptions) {
+	t := val.Type()
+
+	// Also do nothing if this is the type's zero value
+	if t.Comparable() && val.Interface() == reflect.Zero(t).Interface() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Array, reflect.Slice:
+		// formatName automatically adds square brackets, so just pass an empty
+		// string into the breadcrumb trail
+		arrNames := append(names, "")
+
+		for i := 0; i < val.Len(); i++ {
+			// The one exception to the above is when options have requested
+			// that this array/slice be indexed. In that case we produce a hash
+			// keyed with integers which the Stripe API knows how to interpret.
+			if options != nil && options.IndexedArray {
+				arrNames = append(names, strconv.Itoa(i))
+			}
+
+			reflectValue(values, val.Index(i), arrNames, nil)
+		}
+
+	case reflect.Bool:
+		if options != nil {
+			if val.Bool() {
+				switch {
+				case options.Empty:
+					values.Add(formatName(names), "")
+				case options.Invert:
+					values.Add(formatName(names), strconv.FormatBool(false))
+				case options.Zero:
+					values.Add(formatName(names), "0")
+				}
+			}
+		} else {
+			values.Add(formatName(names), strconv.FormatBool(val.Bool()))
+		}
+
+	case reflect.Float32:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 32))
+
+	case reflect.Float64:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 64))
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
```

Comment:
Yeah we should do that (we currently have a mix of all kinds of `int`, `int64` and `uint64` when basically everything could just be `int`), but lets try to establish that separately. It seems like just good hygiene for our encoder here to support all possible types that could be reflected.